### PR TITLE
*: check tiflash peer status for update replica available value

### LINKED ISSUE #35914: Need update TiFlash replica available value when some of regions is down or pending.
## Enhancement
- If some of the regions in TiFlash sync very slow.
- Then in TiFlash, The current peer may report itself as a `down-peer` or a `pending-peer`
- On the TiDB side, it still will see the TiFlash replica available as `true`. So the query may still send to the TiFlash.
- But in this time, lots of regions have not be synced, and The query in TiFlash will got `region timeout error`

### PR DESCRIPTION
**Pull request by jiaqizho**
### What problem does this PR solve?

Issue Number: close #35914

Problem Summary:

### What is changed and how it works?
- Add a poll method named `pollTiFlashPeerInfo` in `ddl_tiflash_api.go`
- `pollTiFlashPeerInfo` will request the PD API `regions/check/down-peer|pending-peer`, Confirm whether the tables(which set TiFlash replicated > 1) can still provide services through the invalid peers(down|pending peers) corresponding to the returned regions.
- If table is unavailable, Then generate a DDL to update the table replicate available.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: domain/infosync/tiflash_manager.go [hunk 1]
```diff
@@ -405,6 +442,58 @@ func (tiflash *MockTiFlash) HandleGetPDRegionRecordStats(_ int64) helper.PDRegio
 	}
 }
 
+// HandleGetPDInvalidRegionsInfo is mock function for GetPDInvalidRegionsInfo.
+// It currently always returns 1 region with 1 down-peer Region and 1 pending peer
```
Comment (flowbehappy) [inline domain/infosync/tiflash_manager.go:446]: Why is it a mock function?

Comment (jiaqizho) [reply domain/infosync/tiflash_manager.go:447]: The interface in `MockTiFlash ` must be mock function.
If we define a function in `tiflash_manager` then need define a mock function in here.

Comment (flowbehappy) [reply domain/infosync/tiflash_manager.go:447]: ok

File: ddl/ddl_tiflash_api.go [hunk 2]
```diff
@@ -463,6 +511,125 @@ func (d *ddl) pollTiFlashReplicaStatus(ctx sessionctx.Context, pollTiFlashContex
...
+	// Collect TiFlash Replica info, for every table.
+	for _, db := range schema.AllSchemas() {
+		tbls := schema.SchemaTables(db.Name)
+		for _, tbl := range tbls {
+			tblInfo := tbl.Meta()
+			GetTiFlashReplicaMapInfo(tblInfo, &tableMap)
+		}
+	}
+
+	tiflashStoreIDs := make([]int64, len(pollTiFlashContext.TiFlashStores))
```
Comment (flowbehappy) [inline ddl/ddl_tiflash_api.go:537]: How about using a map?

File: ddl/ddl_tiflash_api.go [hunk 3]
```diff
@@ -463,6 +511,125 @@ func (d *ddl) pollTiFlashReplicaStatus(ctx sessionctx.Context, pollTiFlashContex
...
+				temp[item] = struct{}{}
+				result = append(result, item)
+			}
+		}
+		return result
+	}
+
+	var invaildRegions helper.RegionsInfo
+
+	if err := infosync.GetTiFlashPDInvalidRegionsInfo(context.Background(), &invaildRegions); err != nil {
```
Comment (flowbehappy) [inline ddl/ddl_tiflash_api.go:574]: I'm afraid that will be a lot of regions.

File: ddl/ddl_tiflash_api.go [hunk 4]
```diff
@@ -463,6 +511,125 @@ func (d *ddl) pollTiFlashReplicaStatus(ctx sessionctx.Context, pollTiFlashContex
...
+		}
+
+		if len(invalidPeerIDs) != 0 {
+			sort.Slice(invalidPeerIDs, func(i, j int) bool { return invalidPeerIDs[i] < invalidPeerIDs[j] })
+
+			invalidPeerIDs = removeDuplicate(invalidPeerIDs)
+			if isTiflashAllPeerDown(tiflashStoreIDs, invalidPeerIDs) {
+				invaildTableID := helper.GetTiFlashTableIDFromEndKey(invaildRegion.EndKey)
+				invalidTableIDs = append(invalidTableIDs, invaildTableID)
+			}
```
Comment (flowbehappy) [inline ddl/ddl_tiflash_api.go:604]: I think we can use an `invalidTableMap`. If we found any region of a table is invalid, then we can skip all regions of the table.

File: ddl/ddl_tiflash_api.go (hunk 3 above)
Comment (flowbehappy) [reply ddl/ddl_tiflash_api.go:574]: I wonder whether we can do this check on PD, instead of TiDB.

File: ddl/ddl_tiflash_api.go (hunk 4 above)
Comment (jiaqizho) [reply ddl/ddl_tiflash_api.go:604]: I don't get it. cause we always need loop all of `invaildRegions` to compare with all of the `tableIDs`.

File: ddl/ddl_tiflash_api.go (hunk 3 above)
Comment (jiaqizho) [reply ddl/ddl_tiflash_api.go:574]: Because the "down|pending" peer info, TIDB won't store it, need request to PD ask for it.

Comment (jiaqizho) [reply ddl/ddl_tiflash_api.go:574]: If we do this job in PD, PD can't see any `schema` or `table` info in that. we also can't get any info about current `TiFlash peer info`.

File: ddl/ddl_tiflash_api.go (hunk 4 above)
Comment (flowbehappy) [reply ddl/ddl_tiflash_api.go:604]: for each region with invalid status
  if (the table id of the region contained by `invalidTableMap`)
    next region
  if (the region is in by a table with tiflash replica && the region's tiflash peer is all invalid)
    add the table id to `invalidTableMap`
end

Comment (jiaqizho) [reply ddl/ddl_tiflash_api.go]: Changed this part of logic.

File: ddl/ddl_tiflash_api.go (hunk 2 above)
Comment (jiaqizho) [reply ddl/ddl_tiflash_api.go:538]: it used to grep the invaildTableIDs..

File: ddl/ddl_tiflash_api.go [hunk 5]
```diff
@@ -463,6 +512,128 @@ func (d *ddl) pollTiFlashReplicaStatus(ctx sessionctx.Context, pollTiFlashContex
...
+		for invalidTableID, invalidPeerIDs := range invalidTableWithPeerIDs {
+			sort.Slice(invalidPeerIDs, func(i, j int) bool { return invalidPeerIDs[i] < invalidPeerIDs[j] })
+			invalidPeerIDs = removeDuplicate(invalidPeerIDs)
+			if isTiflashAllPeerDown(tiflashStoreIDs, invalidPeerIDs) {
+				invalidTableIDs = append(invalidTableIDs, invalidTableID)
+			}
+		}
+	}
+
+	invalidTableIDs = removeDuplicate(invalidTableIDs)
```
Comment (flowbehappy) [inline ddl/ddl_tiflash_api.go:611]: Why not use a map?

File: ddl/ddl_tiflash_api.go [hunk 6]
```diff
@@ -287,6 +293,49 @@ func GetTiFlashReplicaInfo(tblInfo *model.TableInfo, tableList *[]TiFlashReplica
 	}
 }
 
+// GetTiFlashReplicaMapInfo parses tableID + model.TableInfo into map[int64]TiFlashReplicaStatus
+func GetTiFlashReplicaMapInfo(tblInfo *model.TableInfo, tableMap *map[int64]TiFlashReplicaStatus) {
```
Comment (breezewish) [inline ddl/ddl_tiflash_api.go:297]: Actually `tableMap` can be `map[int64]TiFlashReplicaStatus`.

See https://go.dev/play/p/pb3VZdLLJOo

```go
package main

import "fmt"

func foo(v map[int]int) {
	v[1] = 2
}

func main() {
	x := map[int]int{}
	foo(x)
	fmt.Println(x[1])
}
```

BTW, maybe it's better to indicate that this function write result into something. How about naming it as:

```go
func writeTiFlashReplicaInfo(tblInfo *model.TableInfo, intoMap map[int64]TiFlashReplicaStatus)
```

File: ddl/ddl_tiflash_api.go [hunk 7]
```diff
@@ -287,6 +293,49 @@ func GetTiFlashReplicaInfo(tblInfo *model.TableInfo, tableList *[]TiFlashReplica
...
+	if tblInfo.TiFlashReplica == nil {
+		// reject tables that has no tiflash replica such like `INFORMATION_SCHEMA`
+		return
+	}
+	if pi := tblInfo.GetPartitionInfo(); pi != nil {
+		for _, p := range pi.Definitions {
+			logutil.BgLogger().Debug(fmt.Sprintf("Table %v has partition %v\n", tblInfo.ID, p.ID))
```
Comment (breezewish) [inline ddl/ddl_tiflash_api.go:304]: Please remove these debug logs. They are too verbose.

File: ddl/ddl_tiflash_api.go [hunk 8]
```diff
@@ -463,6 +512,126 @@ func (d *ddl) pollTiFlashReplicaStatus(ctx sessionctx.Context, pollTiFlashContex
...
+
+	var tableMap = make(map[int64]TiFlashReplicaStatus)
+	// Collect TiFlash Replica info, for every table.
+	for _, db := range schema.AllSchemas() {
+		tbls := schema.SchemaTables(db.Name)
+		for _, tbl := range tbls {
+			tblInfo := tbl.Meta()
+			GetTiFlashReplicaMapInfo(tblInfo, &tableMap)
+		}
+	}
```
Comment (breezewish) [inline ddl/ddl_tiflash_api.go:536]: @crazycs520 Are there better ways to iterate all tables?

File: ddl/ddl_tiflash_api.go [hunk 9]
```diff
@@ -463,6 +512,126 @@ func (d *ddl) pollTiFlashReplicaStatus(ctx sessionctx.Context, pollTiFlashContex
...
+
+	tiflashStoreIDs := make([]int64, len(pollTiFlashContext.TiFlashStores))
+
+	for i, store := range pollTiFlashContext.TiFlashStores {
+		tiflashStoreIDs[i-1] = store.Store.ID
```
Comment (breezewish) [inline ddl/ddl_tiflash_api.go:541]: Why is it `i-1`?

File: ddl/ddl_tiflash_api.go [hunk 10]
```diff
@@ -463,6 +512,126 @@ func (d *ddl) pollTiFlashReplicaStatus(ctx sessionctx.Context, pollTiFlashContex
...
+	}
+
+	sort.Slice(tiflashStoreIDs, func(i, j int) bool { return tiflashStoreIDs[i] < tiflashStoreIDs[j] })
+
+	isTiflashPeer := func(tiflashPeerIDs []int64, peerID int64) bool {
```
Comment (breezewish) [inline ddl/ddl_tiflash_api.go:546]: PeerID and StoreID are different IDs in PD. Maybe you should correct your naming to store ID.

File: ddl/ddl_tiflash_api.go [hunk 11]
```diff
@@ -463,6 +512,126 @@ func (d *ddl) pollTiFlashReplicaStatus(ctx sessionctx.Context, pollTiFlashContex
...
+		for _, tiflashPeerID := range tiflashPeerIDs {
+			if tiflashPeerID == peerID {
+				return true
+			}
+		}
```
Comment (breezewish) [inline ddl/ddl_tiflash_api.go:551]: As this is sorted, how about using binary search?

File: ddl/ddl_tiflash_api.go [hunk 12]
```diff
@@ -463,6 +512,126 @@ func (d *ddl) pollTiFlashReplicaStatus(ctx sessionctx.Context, pollTiFlashContex
...
+
+		return false
+	}
+
+	isTiflashAllPeerDown := func(tiflashPeerIDs []int64, tiflashDownPeerIDs []int64) bool {
```
Comment (breezewish) [inline ddl/ddl_tiflash_api.go:556]: Seems that this requires that both array are sorted. Is it true?

File: ddl/ddl_tiflash_api.go (hunk 2 above)
Comment (breezewish) [reply ddl/ddl_tiflash_api.go:538]: Maybe you can also build a map for `invalidPeerIDs`, and then check whether the two map are equal (length is identical && everything in map0 is in map1).

**Review by breezewish — COMMENTED** [review]
I have some concerns about the performance of checking peers.

File: domain/infosync/tiflash_manager.go [hunk 13]
```diff
@@ -168,6 +170,43 @@ func (m *TiFlashPDPlacementManager) GetPDRegionRecordStats(ctx context.Context,
...
+		return fmt.Errorf("TiFlashPDPlacementManager returns error in GetPDRegionRecordStats")
+	}
+
+	err = json.Unmarshal(res, regionInfo)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	regionInfo = (*regionInfo).Merge(&tempRegionInfo)
+	logutil.BgLogger().Info("Got down|pending", zap.Int64("region counts", regionInfo.Count))
```
[the remark this answers is not in the document]
Comment (jiaqizho) [reply domain/infosync/tiflash_manager.go:206]: two mins one round. i think it's ok.

File: ddl/ddl_tiflash_api.go (hunk 6 above)
Comment (jiaqizho) [reply ddl/ddl_tiflash_api.go:319]: just keep the style with `GetTiFlashReplicaInfo`

File: ddl/ddl_tiflash_api.go (hunk 12 above)
Comment (jiaqizho) [reply ddl/ddl_tiflash_api.go:662]: both sorted

**Comment by flowbehappy** [comment]
/hold 
Hold this PR until we support batch DDL on `ActionUpdateTiFlashReplicaStatus`, otherwise this PR could cause too many DDL operations in the case of a large amount of tables.

File: parser/model/model.go [hunk 14]
```diff
@@ -602,6 +602,8 @@ func (t TiFlashMode) String() string {
 type TiFlashReplicaInfo struct {
 	Count                 uint64
 	LocationLabels        []string
+	Ready                 bool
```
Comment (lcwangchao) [inline parser/model/model.go:605]: Can you add some comments for `Ready` and `Available` ? These two states are confusing.

File: ddl/ddl_tiflash_api.go [hunk 15]
```diff
@@ -463,6 +544,198 @@ func (d *ddl) pollTiFlashReplicaStatus(ctx sessionctx.Context, pollTiFlashContex
...
+		tiflashStoreIDs[loopIndex] = store.Store.ID
+		loopIndex++
+	}
+
+	sort.Slice(tiflashStoreIDs, func(i, j int) bool { return tiflashStoreIDs[i] < tiflashStoreIDs[j] })
+
+	isTiflashPeer := func(tiflashPeerIDs []int64, peerID int64) bool {
+		left := 0
+		right := len(tiflashPeerIDs)
+		for left <= right {
```
[the remark this answers is not in the document]
Comment (jiaqizho) [reply ddl/ddl_tiflash_api.go:654]: https://pkg.go.dev/sort#Search
seem like it does not support `int64`?

**Pull request closed without merging**